src/abstractdataframe/abstractdataframe.jl

```diff
@@ -2525,39 +2525,95 @@ end
 
 function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.invpermute!!)},
                               df::AbstractDataFrame, p::AbstractVector{<:Integer})
-    toskip = Set{Int}()
+    length(p) == nrow(df) || throw(ArgumentError("Permutation does not have a correct length"))
+    
+    cp = _compile_permutation!(Base.copymutable(p))
+
+    isempty(cp) && return df
+
+    if fun === Base.invpermute!! 
+        pop!(cp)
+        reverse!(cp)
+        push!(cp, 0)
```

Comment:
Interestingly it is roughly equally fast as `reverse!(@view cp[1:end-1])`

Reply:
Great! I've verified that with a microbenchmark that finds `reverse!` slightly faster. I'm guessing the compiler makes the view transparent. I pushed that change.

Reply:
Thank you (I have not pushed for this as the difference was minimal 😄)